File: microservice-architecture/cdc-pattern/python/test/test_cdc.py
import os
import sys
import unittest

sys.path.append(os.getcwd())

from src.cdc import Broker, DataStore  # noqa: E402


class CDCPatternTest(unittest.TestCase):
    def test_cdc_pattern(self) -> None:
        store = DataStore()
        broker = Broker()
        store.create_order("ORD-1001")
        self.assertFalse(store.changes[0].processed)
        store.relay_changes(broker)
        self.assertEqual(["CHG-ORD-1001"], broker.published)
        self.assertTrue(store.changes[0].processed)
        store.relay_changes(broker)
        self.assertEqual(["CHG-ORD-1001"], broker.published)


if __name__ == "__main__":
    unittest.main()

File: microservice-architecture/cdc-pattern/js/test/test_cdc.js
import { Broker, DataStore } from '../src/cdc.js'

function assertEqual(expected, actual, message) {
  if (expected !== actual) throw new Error(`${message} expected=${expected} actual=${actual}`)
}

const store = new DataStore()
const broker = new Broker()
store.createOrder('ORD-1001')
assertEqual(false, store.changes[0].processed, 'initial')
store.relayChanges(broker)
assertEqual(1, broker.published.length, 'publish once')
assertEqual(true, store.changes[0].processed, 'processed')
store.relayChanges(broker)
assertEqual(1, broker.published.length, 'no duplicate')
console.log('cdc-pattern(js) tests passed')
